Change transaction queue limit and broadcast policy to be fee based

MonsieurNicolas opened the pull request:
# Description
This PR changes how we manage the transaction queue limit.
* Main change is that we move away from a hard limit (banning transactions when the cap is hit), and instead use fee bids as a way to evict lower fee transactions.
* The second change builds on the new policy and adds a mechanism to lock out all transactions that are low bids until the transaction queue reaches a lower utilization. This should allow to de-prioritize further spikes of low fee transactions.

Opening this  in draft to get an initial round of feedback.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

graydon reviewed (changes requested):
Minor tweaks inline -- weakly-approved if addressed.

But broader concern is this feels to me like it's pushing against limits of understandability of this subsystem. In particular, two things worry me:

1. The back-and-forth chatter around deleting elements seems fairly subtle and reentrant -- asking for a container or iterator invalidation (leading to memory fault) at some point. This call tree worries me:

```
TxQ::tryAdd
  TxQ::canAdd
    if !TxQLim::canAddTx
      TxQ::ban
        TxQ::dropTransactions
          TxQ::prepareDropTransaction
           TxQLim::removeTransaction
  TxQLim::evictTransactions
    TxQ::ban
      TxQ::dropTransactions
        TxQ::prepareDropTransaction
         TxQLim::removeTransaction
```

2. I find it fairly challenging to picture the evolution and interaction of the two layers of policy here: admission control based on decisions made up in TxQ::{mAccountStates,mBannedTransactions} and secondary admission control based on the fee groups down in TxQLimiter::mTxGroupsByBaseFee. Maybe that's just "how we have to do it", but it seems tricky to even extract a description of what the policy _is_ much less convey it to users in a way they'll be able to reason about.

If I take a step back, my _feeling_ is that the whole TxQ subsystem is fundamentally just .. one central set-of-txs (partially ordered by causality/seqno), plus some auxiliary indexes, counters and aggregate values to support admission-control, replacement and broadcast decisions; and it's being made quite complicated here by the need for all that auxiliary information to be stored in summary-structures that can be updated incrementally as txs are admitted or rejected.

Reading I can't help but wonder if there's any possibility of making the underlying nature simpler and more obvious, and making the derived indexes and aggregates be recomputed either on-demand, or using batched recomputation every N amitted txs or something. We're only talking about a few thousand elements evolving over seconds of real time anyways; it's not like it's millions or billions of elements. The quest for O(1) everything here seems to be really working against comprehensibility.

MonsieurNicolas commented on src/herder/TransactionQueue.cpp:
```diff
@@ -206,7 +204,7 @@ TransactionQueue::canAdd(TransactionFrameBasePtr tx,
     }
 
-    if (netOps + mQueueSizeOps > maxQueueSizeOps())
+    if (mLimiter.canAddTx(tx, oldTx))
```
ah that was a refactor bug sorry

MonsieurNicolas replied:
fixed

MonsieurNicolas commented:
> If I take a step back, my feeling is that the whole TxQ subsystem is fundamentally just .. one central set-of-txs (partially ordered by causality/seqno), plus some auxiliary indexes, counters and aggregate values to support admission-control, replacement and broadcast decisions; and it's being made quite complicated here by the need for all that auxiliary information to be stored in summary-structures that can be updated incrementally as txs are admitted or rejected

I found that separating out concerns, while it might bit less efficient from a memory point of view, ends up with much simpler tests:
the issue here is that tracking all those things inside one big data structure inside `TransactionQueue` just makes testing quite hard. I spent a full day trying to reconcile the "fee bump transaction tests" with limits, and just gave up as tests were becoming extremely complicated and coverage was actually lower than what is proposed in this PR.

I think that in a subsequent PR I want to actually do the same thing with the broadcasting code: I'd prefer to keep the management of sequence numbers in its own dedicated box as it's the main source of complexity in this code.

> Reading I can't help but wonder if there's any possibility of making the underlying nature simpler and more obvious, and making the derived indexes and aggregates be recomputed either on-demand, or using batched recomputation every N amitted txs or something. We're only talking about a few thousand elements evolving over seconds of real time anyways; it's not like it's millions or billions of elements. The quest for O(1) everything here seems to be really working against comprehensibility.

That is something I considered (recompute the whole thing all the time - I would not consider doing it only every "N" as it becomes too hard to reason about what the thing is actually doing), but chickened out because
* this logic is called in the order of 1M times per ledger (while transactions get flooded),
* per my earlier response, as I want to separate the broadcast logic, I'll need some state persisted. So we're back to needing incremental changes.

graydon commented:
Fair points, it was a vague-unease comment to begin with. However:

> this logic is called in the order of 1M times per ledger (while transactions get flooded)

This statement is interesting to me! We have duplicate flood-packet suppression code in the overlay receive path; I wouldn't think txs should be hitting this code anywhere near this often. Are you sure?

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,171 @@
+uint64
+getIDFromAccountID(AccountID const& accountID)
+{
+    return stellar::shortHash::computeHash(stellar::ByteSlice(
```
Maybe it's overkill I agree, I can remove it as it's only for a few thousand transactions.
Reason here (not clear because I forgot to move some code, read on) is we potentially have a "walk all transactions" degenerate case in `canAddTx` (all transactions have a different fee), and we're looking up the `id` in all fee groups (note: the *computation* of the id should really be done outside the loop).

I am realizing that I forgot to use the `UnorderedSet/Map` templates instead of the `std` versions, so I need to touch that code anyways

MonsieurNicolas replied:
removed and moved to proper types

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,171 @@
+        throw std::logic_error(
+            "invalid state getting worst tx in TxQueueLimiter");
+    }
+    // as we hash pointers, picking the first one in hash ordeer is
```
fixed

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -5,16 +5,22 @@
+    , mBaseFeeBidMultiplier(feeMultiplier)
     , mLedgerManager(lm)
 {
+    if (mBaseFeeBidMultiplier < 1)
```
Either the condition is wrong or the exception is wrong. Do you want to accept 1? If so, the error message should probably say "Multiplier must be at least 1"

You might want to check the same for `mPoolLedgerMultiplier`.

jonjove reviewed (changes requested):
I reviewed everything except for the tests and the last commit (because I think there are going to be many changes before that point, so better to just wait). Some comments point out issues that were fixed in later commits. I left them in for clarity, but feel free to resolve those.

MonsieurNicolas commented on src/herder/TransactionQueue.h:
```diff
@@ -118,6 +119,7 @@ class TransactionQueue
 
     explicit TransactionQueue(Application& app, int pendingDepth, int banDepth,
                               int poolLedgerMultiplier);
+    ~TransactionQueue();
```
Without this, Herder also needs the definition for `TxQueueLimiter` (I think it tries to inline the destructor?) and I'm trying to avoid bloating the include hierarchy with this stuff

jonjove replied:
Yes, that's right. Hard to tell by looking at the code though (even as a PR review where I can see this was added alongside the `unique_ptr` I didn't notice, it'll be even harder when maintaining the code in the future), so please add a comment.

MonsieurNicolas commented on src/herder/test/TransactionQueueTests.cpp:
```diff
@@ -223,76 +228,54 @@ TEST_CASE("TransactionQueue", "[herder][transactionqueue]")
     auto account3 = root.create("a3", minBalance2);
 
-    auto txSeqA1T0 = transaction(*app, account1, 0, 1, 100);
+    auto txSeqA1T0 = transaction(*app, account1, 0, 1, 200);
```
I want the base fee to be the same (200) for all txs

jonjove replied:
The test passes (at least as of this commit) either way. Is this to make the test work for some later commit?

MonsieurNicolas replied:
no, I did this as I was trying to understand what the test cases were doing.

Net is: this test section does *not* care about fees, but the setup made it look like they did (each transaction had a higher fee than the previous one) so it was confusing. What is important in those txs are sequence numbers and amounts.

Tests that care about fees are in the "replace by fee" group or "TransactionQueue limits".

MonsieurNicolas commented on src/herder/TransactionQueue.cpp:
```diff
@@ -302,7 +300,7 @@ TransactionQueue::tryAdd(TransactionFrameBasePtr tx)
     stateIter->second.mQueueSizeOps += ops;
     stateIter->second.mBroadcastQueueOps += ops;
-    mQueueSizeOps += ops;
+    mLimiter.removedTransaction(tx);
```
Sorry, I did the same thing to Graydon, this is fixed in the next commit.

That first commit was true garbage, probably created it way too early: I had all possible typos and bugs possible.
I'll fixup this commit to avoid further confusion for a future observer.

MonsieurNicolas commented on src/herder/TransactionQueue.cpp:
```diff
@@ -930,4 +920,47 @@ TransactionQueue::isFiltered(TransactionFrameBasePtr tx) const
+void
+TransactionQueue::Limiter::addedTransaction(TransactionFrameBasePtr& tx)
+{
+    mQueueSizeOps += tx->getNumOperations();
```
I added a check in the next commit. I was trying to keep the "refactor" commit as close to the previous behavior as possible

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,160 @@
+            return false;
+        }
+        auto evictTx = getWorstTransaction();
+        evict(evictTx);
```
I cannot do this, see the contract in the header file: this is the responsibility of `evict` (which it does by calling `ban`). The lifetime of transactions is determined by `TransactionQueue`.

If I was calling `removeTransaction` here I would cause this class and the main `TransactionQueue` to be out of sync, and `evict` would end up calling `removeTransaction` again which will trip invariants/violate any lifetime expectations on which transactions should be referenced at any given time by the different classes.

jonjove replied:
Didn't read the comment. This is fine since it's documented.

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -157,4 +174,10 @@ TxQueueLimiter::reset()
     mTxGroupsByBaseFee.clear();
     mQueueSizeOps = 0;
```
good catch

MonsieurNicolas commented on src/herder/TxQueueLimiter.h:
```diff
@@ -0,0 +1,62 @@
+    int const mPoolLedgerMultiplier;
+    LedgerManager& mLedgerManager;
+
+    struct TxGroup
```
You might be right - when I originally  wrote the code, I thought that the worst case for `canAdd` was walking the entire set of transactions, but really the worst case is walking the worst 101 transactions, and even performing the extra comparison against the source account for all those transactions is not a lot of work.

MonsieurNicolas replied:
reworked it as per your recommendation - a lot less code you will see

MonsieurNicolas commented:
Updated the PR:

* addressed (I think) all feedback from @jonjove
    * reworked limit code to just use a set instead of the nested maps (negligible perf hit, a lot less code)
* changed the delayed flooding code to prioritize transactions based on fee
   * last few commits are changing the defaults to something more sensible as well as making delayed flooding the default

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -80,6 +83,15 @@ TxQueueLimiter::removeTransaction(TransactionFrameBasePtr& tx)
+    if (mMinBaseFee != 0)
+    {
+        auto curFree = 100 - (mQueueSizeOps * 100) / maxQueueSizeOps();
+        if (curFree >= mThresholdFreePercentUnstick)
```
I ended up removing this code for now, reset happens when ledgers gets closed

jonjove commented on src/herder/TransactionQueue.cpp:
```diff
@@ -737,98 +743,116 @@ TransactionQueue::broadcastTx(AccountState& state, TimestampedTx& tx)
         {
-            iters.emplace_back(
-                TxTracker{m.second.mTransactions.begin(), &m.second});
+            iters.emplace(
```
This _almost_ produces the right broadcast order, but not exactly. There is one error. Example:
- `mAccountStates.size() == 2` so call the first element `m1` and the second `m2`
-  `m1` has 2 transactions each with 1 operation. The first has `fee = 100` and the second has `fee = 200` (the transaction queue limiter permits this as long as adding the higher fee wouldn't have required the lower fee to be removed). The first one has been broadcasted but the second one hasn't.
- `m2` has 1 transaction with 1 operation. It has `fee = 150` and has not been broadcasted (this is possible if it was added after the first `m1` transaction had already been broadcasted).
- The next transaction that should be broadcasted is the second `m1` transaction because it has the highest fee of all remaining transactions. But `m2` will be at the top of the priority queue because the `m2` transaction has a higher fee than the first `m1` transaction. Therefore the `m2` transaction will broadcast before the second `m1` transaction which is not in surge pricing order.

The `while` loop on 847,851 needs to be duplicated here. Note that the `while` loop still is needed at 847,851 to properly handle fee-bump replace-by-fee in the middle of a queue; otherwise the same situation described above can happen in that case. Please add a comment there to clarify why the loop is needed in both places. Furthermore, because the loop is being used twice I recommend that you factor it into a function.

MonsieurNicolas replied:
yes indeed great catch - a very subtle fairness issue

jonjove commented on docs/stellar-core_example.cfg:
```diff
@@ -138,11 +138,11 @@ MAX_BATCH_WRITE_COUNT=1024
 # If negative, the number is derived as:
 #  -FLOOD_OP_RATE_PER_LEDGER*<maximum number of operations per ledger>
-FLOOD_OP_RATE_PER_LEDGER = -2
+FLOOD_OP_RATE_PER_LEDGER = -1.0
```
Maybe we should just eliminate the possibility of "fixed number" flooding (eg. if this is positive) and always make it multiplicative? It's really strange to configure the fixed number as a floating point.

MonsieurNicolas replied:
agree, removed the fixed version that we don't need

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -103,11 +103,14 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
     if (txBaseFee <= getMinBaseFee())
     {
-        return false;
+        auto minFee = newTx->getFee(Config::CURRENT_LEDGER_PROTOCOL_VERSION,
```
Nitpick: The formatting on this is bizarre. Did `clang-format` force this?

MonsieurNicolas replied:
deleted

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -103,11 +103,14 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
     {
-        return false;
+        auto minFee = newTx->getFee(Config::CURRENT_LEDGER_PROTOCOL_VERSION,
+                                    txBaseFee, false) +
```
This doesn't seem right, shouldn't it say `getMinBaseFee()` here?

MonsieurNicolas replied:
removed

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -103,11 +103,14 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
     if (txBaseFee <= getMinBaseFee())
     {
-        return false;
+        auto minFee = newTx->getFee(Config::CURRENT_LEDGER_PROTOCOL_VERSION,
```
Nitpick: Calling `getFee` is a little overkill here--I had to go check why we were using that function specifically. I would just do the multiplication explicitly here (after all, that's what happening inside of `getFee`) but it will avoid passing the version number, which is ignored anyway because you are passing `false` as the last argument.

MonsieurNicolas replied:
got rid of that

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -117,22 +120,29 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
-            return false;
+            int64 minFee;
+            // compute a fee strictly better than `it`
+            // this won't overflow because fee bids for valid transactions are
```
This comment is wrong. This calculation happens _before_ the available balance check, and fee-bump transactions have `int64` fees. So this can overflow, and you need to be more careful.

MonsieurNicolas replied:
updated the code to not care if we overflow

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -117,22 +120,29 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
         neededOps -= curOps;
     }
-    return false;
+    return std::make_pair(false, 0ll);
```
This case is interesting... it means that the transaction just didn't fit in the queue even if you remove everything else? Or am I missing something? If I'm reading this correctly, then we should probably add a comment here that explains what this means. And we should probably return a special value which means something like "submit to another node".

MonsieurNicolas replied:
added a comment to clarify

MonsieurNicolas replied:
I don't think we should add a new return value: it's just a variation on "try again later" type of situation for the client

jonjove replied:
Well, it's a bit different. If you get here, you will _never_ succeed no matter how long you wait unless the node changes configuration. Basically the way I see it is, if you get here then you are totally broken (eg. send an alert to operations) whereas if you hit any of the other cases then you can just back off and wait.

MonsieurNicolas replied:
OK for operators, this makes sense. I am adding a warning when the limiter is configured that way.

jonjove reviewed (changes requested):
Looks a lot better now, but I think I still found two more bugs. We'll need to go through at least one more review cycle.

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,175 @@
+            return cmp3 < 0;
+        }
+        // break tie with pointer arithmetic
+        auto lx = reinterpret_cast<size_t>(l.get()) ^ mSeed;
```
That's not the responsibility of this class and reduces coupling: it doesn't care about sequence numbers (proper sequencing, including duplicates) either.

jonjove replied:
I'm just proposing defense in depth. I don't think the consequences of having two identical transactions in here would be critical anyway, so I'll resolve this.

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,175 @@
+            return false;
+        }
+        // ensure that this transaction is not from the same account
+        auto& tx = *it;
```
it's already deduced like this

MonsieurNicolas commented on src/transactions/TransactionFrameBase.cpp:
```diff
@@ -24,4 +24,10 @@ TransactionFrameBase::makeTransactionFromWire(Hash const& networkID,
         abort();
     }
 }
+int64_t
```
fixed

MonsieurNicolas commented on src/transactions/TransactionFrameBase.cpp:
```diff
@@ -24,4 +24,10 @@ TransactionFrameBase::makeTransactionFromWire(Hash const& networkID,
+int64_t
+TransactionFrameBase::getBaseFeeBid() const
+{
+    return bigDivide(getFeeBid(), 1, static_cast<int64_t>(getNumOperations()),
```
I got rid of the dependency on base fee bid and instead only use 3way compare everywhere

MonsieurNicolas commented on src/transactions/TransactionFrameBase.cpp:
```diff
@@ -24,4 +24,10 @@ TransactionFrameBase::makeTransactionFromWire(Hash const& networkID,
     }
 }
+int64_t
+TransactionFrameBase::getBaseFeeBid() const
```
deleted

MonsieurNicolas commented on src/herder/TransactionQueue.cpp:
```diff
@@ -930,4 +920,47 @@ TransactionQueue::isFiltered(TransactionFrameBasePtr tx) const
+TransactionQueue::Limiter::canAddTx(TransactionFrameBasePtr& tx,
+                                    TransactionFrameBasePtr& oldTx)
+{
+    auto netOps = tx->getNumOperations();
```
added an invariant check: it would be a bug to replace a transaction with less operations in the current design

jonjove replied:
I don't think this is right. Fee-bump replace-by-fee can cause a smaller transaction to replace a bigger transaction (the inner transactions do not need to have the same hash). So unless I'm missing something, this breaks fee-bump replace-by-fee.

This revealed to me that we don't test that case right now (eg. use fee-bump to replace 10 operations with 1 operation) so please add a test case for that since it is relevant to these changes.

jonjove replied:
Never mind, you reworked this in a later commit but I responded before I reached that change. I think the new approach is fine. But it is still a good idea to test replacing a bigger transaction with a smaller one, because that is in fact possible.

MonsieurNicolas replied:
my comment in the PR was stale as well: I realized later that we the only assumption in the limiter is that "old" is already in the queue (it's a replace by X situation). The comment and invariant in the code should be the right one now.

jonjove replied:
Adding additional tests is tracked in https://github.com/stellar/stellar-core/issues/2913

MonsieurNicolas commented on src/herder/TransactionQueue.h:
```diff
@@ -204,21 +204,45 @@ class TransactionQueue
+        }
+        size_t maxQueueSizeOps() const;
+
+        void addedTransaction(TransactionFrameBasePtr& tx);
```
fixed

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,175 @@
+              mComparator)
+    {
+    }
+    QueueLimiterTxMap(QueueLimiterTxMap const&) = delete;
```
remembered we have helpers for this - fixed

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,175 @@
+    }
+};
+
+class QueueLimiterTxMap
```
Yeah I found this the hard way in a different PR: when copying containers, comparators do not get copied and instead get instantiated by type so a copy would actually behave differently.

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,175 @@
+    : public std::set<TransactionFrameBasePtr, QueueLimiterTxComparator>
+{
+  public:
+    QueueLimiterTxComparator mComparator;
```
yep

MonsieurNicolas commented on src/herder/TxQueueLimiter.h:
```diff
@@ -0,0 +1,56 @@
+    LedgerManager& mLedgerManager;
+
+    // all known transactions
+    std::unique_ptr<QueueLimiterTxMap> mTxs;
```
It's not so much for hiding (and avoid having too many details in the include file): to implement `reset` I just instantiate a new object

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,175 @@
+TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
+                         TransactionFrameBasePtr& oldTx) const
+{
+    auto newOps = mQueueSizeOps + newTx->getNumOperations();
```
yeah added an invariant check

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -108,6 +108,11 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr& newTx,
     {
         newOps -= oldTx->getNumOperations();
     }
+    auto txBaseFee = newTx->getBaseFeeBid();
```
fixed

MonsieurNicolas commented on src/herder/TransactionQueue.h:
```diff
@@ -210,9 +211,13 @@ class TransactionQueue
 
+    size_t mBroadcastSeed;
+
+    friend struct TxQueueTracker;
```
fixed

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -90,7 +90,26 @@ TxQueueLimiter::removeTransaction(TransactionFrameBasePtr const& tx)
+    int64 minFee = m;
+    int64 v;
+    if (bigDivide(v, refFeeBid, tx->getNumOperations(), refNbOps,
+                  Rounding::ROUND_UP) &&
```
This is a pretty minor detail, but I think this should be `Rounding::ROUND_DOWN`. Consider the following:
```
refFeeBid = 100
tx->getNumOperations() = 2
refNbOps = 3

-> v = ceil(66.66) -> minFee = 68
```
but from fee rate compare
```
201 = 67 * 3 > 100 * 2 = 200
```
so `minFee` should have been 67.

The `v + 1` is still right though, which can be seen from considering the case `100, 1, 1`.

MonsieurNicolas replied:
indeed

jonjove commented on src/herder/TxSetFrame.h:
```diff
@@ -119,7 +119,10 @@ class TxSetFrame : public AbstractTxSetFrameForApply
 
-// 3-way fee rate compare
+// 3-way fee rate compare variants
+int feeRate3WayCompare(int64 lFeeBid, uint32 lNbOps, int64 rFeeBid,
```
Nitpick: these functions are defined in a strange place given that they don't really have anything to do with transaction set frames. I spent a while looking for them because I couldn't remember the name and they weren't where I would have guessed. Maybe put them in TransactionFrameBase.h or something?

MonsieurNicolas replied:
sure done

jonjove commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -110,6 +111,16 @@ TxQueueLimiter::canAddTx(TransactionFrameBasePtr const& newTx,
         newOps -= oldTx->getNumOperations();
     }
     newOps += newTx->getNumOperations();
+    if (mMinFeeNeeded.second != 0)
```
As I commented earlier, the order here is still really weird (and the need for whitespace between different conceptual tasks remains). When I read this code, I am forced to ask myself "why did I compute the number of operations, then switch and evaluate the min fee, then switch back and actually act on the number of operations?" But you don't have to do that, so you shouldn't. Please rearrange so that the function 
1. Checks the min fee
2. Then computes `newOps` and checks if it will fit
3. Then does everything else

It'll be easier to understand for future maintainers.

MonsieurNicolas replied:
yeah agree, this was the result of a convoluted rebase dance. I'll leave this as a separate commit

jonjove reviewed (commented):
We are getting very close here now on the implementation, although I still need to review all the test changes. I have a few more comments (some which were posted already) and I have resolved everything from before that has been entirely addressed.

MonsieurNicolas commented on src/herder/TxQueueLimiter.cpp:
```diff
@@ -0,0 +1,180 @@
+    if (mTxs->erase(tx) == 0)
+    {
+        throw std::logic_error(
+            "invalid state (basefee) removing tx in TxQueueLimiter");
```
me neither - I think it's a leftover from the old implementation. fixed

MonsieurNicolas commented on src/herder/TransactionQueue.cpp:
```diff
@@ -739,98 +746,120 @@ TransactionQueue::broadcastTx(AccountState& state, TimestampedTx& tx)
-        // std::swap ends up calling this operator
-        // See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85828
-        TxTracker&
-        operator=(TxTracker&& other)
```
correct: it was needed because the old code was using `shuffle` that in some cases was calling `std::swap(e, e)`  - now we never move onto self

jonjove commented on src/herder/test/TransactionQueueTests.cpp:
```diff
@@ -649,30 +650,167 @@ TEST_CASE("TransactionQueue", "[herder][transactionqueue]")
+            simpleLimitTest([](int i) { return 100 * (100 - i); });
+        }
+    }
+    SECTION("multi accounts limits")
```
I like the way you wrote this, quite compact and easy to understand.

MonsieurNicolas commented on src/herder/TransactionQueue.cpp:
```diff
@@ -739,98 +746,120 @@ TransactionQueue::broadcastTx(AccountState& state, TimestampedTx& tx)
+{
+    // broadcast transactions in surge pricing order:
+    // loop over transactions by picking from the account queue with the
+    // highgest base fee not broadcasted so far.
```
fixed

jonjove reviewed (commented):
Reviewed the tests too now, that seems good.

jonjove commented on src/util/Math.h:
```diff
@@ -40,4 +40,15 @@ rand_element(std::vector<T> const& v)
+
+template <typename T>
+T&
+rand_element(std::vector<T>& v)
```
Opened https://github.com/stellar/stellar-core/issues/2912.

jonjove commented:
r+ b91efbf

jonjove commented:
r+ a22cf5e
